partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/CastTable.kt

```diff
@@ -0,0 +1,701 @@
+package org.partiql.eval.internal.operator.rex
+
+import com.amazon.ionelement.api.ElementType
+import com.amazon.ionelement.api.IonElementException
+import com.amazon.ionelement.api.createIonElementLoader
+import org.partiql.errors.DataException
+import org.partiql.errors.TypeCheckException
+import org.partiql.eval.value.Datum
+import org.partiql.types.PType
+import java.math.BigDecimal
+import java.math.BigInteger
+import java.math.RoundingMode
+
+@Suppress("DEPRECATION")
+internal object CastTable {
+
+    public fun cast(source: Datum, target: PType): Datum {
+        if (source.isNull) {
+            return Datum.nullValue(target)
+        }
+        if (source.isMissing) {
+            return Datum.missingValue(target)
+        }
+        if (target.kind == PType.Kind.DYNAMIC) {
+            return source
+        }
+        val cast = castTable[source.type.kind.ordinal][target.kind.ordinal]
+            ?: throw TypeCheckException("CAST(${source.type} AS $target) is not supported.")
+        return try {
+            cast.invoke(source, target)
+        } catch (t: Throwable) {
+            throw TypeCheckException("Failed to cast $source to $target")
+        }
+    }
+
+    private val castTable: Array<Array<((Datum, PType) -> Datum)?>> = Array(PType.Kind.values().size) { inputKind ->
```

Comment:
nit: introduce two type aliases

1. alias for the cast function signature
2. alias for the 2d array / table